File: library/camera/src/main/java/com/suishi/camera/utils/OrientationLiveData.java
package com.suishi.camera.utils;

import android.content.Context;
import android.hardware.camera2.CameraCharacteristics;
import android.view.OrientationEventListener;
import android.view.Surface;

import androidx.lifecycle.LiveData;

public class OrientationLiveData extends LiveData<Integer> {


    private Context mContenx;

    private CameraCharacteristics mCharacteristics;

    private OrientationEventListener listener;

    public OrientationLiveData(Context contenx, CameraCharacteristics characteristics) {
        this.mContenx = contenx;
        this.mCharacteristics = characteristics;

        listener=new OrientationEventListener(mContenx.getApplicationContext()) {
            @Override
            public void onOrientationChanged(int orientation) {
                int rotation=0;
                if(rotation<=45) {
                    rotation=Surface.ROTATION_0;
                }else if(rotation<=135) {
                    rotation = Surface.ROTATION_90;
                }else if(rotation<=225) {
                    rotation = Surface.ROTATION_180;
                }else if(rotation<=315) {
                    rotation=Surface.ROTATION_270;
                }else {
                    rotation = Surface.ROTATION_0;
                }
                int relative=computeRelativeRotation(mCharacteristics,rotation);
//                if(relative!=getValue()){
//                    postValue(relative);
//                }
            }
        };
    }

    private static int computeRelativeRotation(CameraCharacteristics characteristics, int surfaceRotation) {
        int sensorOrientationDegrees = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
        int deviceOrientationDegrees;
        switch (surfaceRotation) {
            case Surface.ROTATION_0: {
                deviceOrientationDegrees = 0;
            }
            break;
            case Surface.ROTATION_90: {
                deviceOrientationDegrees = 90;
            }
            break;
            case Surface.ROTATION_180: {
                deviceOrientationDegrees = 180;
            }
            break;
            case Surface.ROTATION_270: {
                deviceOrientationDegrees = 270;
            }
            break;
            default: {
                deviceOrientationDegrees = 0;
            }
        }
        int sign=0;
        if(characteristics.get(CameraCharacteristics.LENS_FACING)==CameraCharacteristics.LENS_FACING_FRONT) {
            sign=1;
        }else {
            sign=-1;
        }
        return (sensorOrientationDegrees-(deviceOrientationDegrees*sign)+360)%360;

    }


    @Override
    protected void onActive() {
        super.onActive();
        listener.enable();
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        listener.disable();
    }
}

File: app/src/main/java/com/suishi/live/app/utils/OrientationLiveData.kt
package com.suishi.live.app.utils

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.view.OrientationEventListener
import android.view.Surface
import androidx.lifecycle.LiveData
import com.seu.magicfilter.utils.Rotation

class OrientationLiveData (context: Context,characteristics: CameraCharacteristics): LiveData<Int>() {

    companion object{

        private fun computeRelativeRotation(characteristics: CameraCharacteristics,surfaceRotation: Int):Int{
            val sensorOrientationDegree=characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION)
            val deviceOrientationDegree=when(surfaceRotation){
                Surface.ROTATION_0-> 0
                Surface.ROTATION_90-> 90
                Surface.ROTATION_180-> 180
                Surface.ROTATION_270 -> 270
                else -> 0
            }

            val sign = if(characteristics.get(CameraCharacteristics.LENS_FACING)==CameraCharacteristics.LENS_FACING_FRONT) 1 else -1
            return (sensorOrientationDegree-(deviceOrientationDegree*sign)+360) %360
        }

    }

    private val listener=object :OrientationEventListener(context.applicationContext){
        override fun onOrientationChanged(orientation: Int) {
            val rotation=when{
                orientation<=45-> Surface.ROTATION_0
                orientation<=135-> Surface.ROTATION_90
                orientation<=225-> Surface.ROTATION_180
                orientation<=315->Surface.ROTATION_270
                else -> Surface.ROTATION_0
            }
            val relative=computeRelativeRotation(characteristics,rotation)
            if(relative!=value) postValue(relative)
        }

    }


    override fun onActive() {
        super.onActive()
        listener.enable()
    }

    override fun onInactive() {
        super.onInactive()
        listener.disable()
    }

}